File: backend/app/models/mahjong.py
from enum import Enum
from typing import List, Optional, Dict, Set
from pydantic import BaseModel, Field


class TileType(Enum):
    """麻将牌类型"""
    WAN = "wan"      # 万
    TIAO = "tiao"    # 条
    TONG = "tong"    # 筒
    ZI = "zi"        # 字牌


class Tile(BaseModel):
    """麻将牌"""
    type: TileType
    value: int = Field(..., ge=1, le=9)  # 1-9
    
    def __str__(self):
        return f"{self.value}{self.type.value}"
    
    @classmethod
    def from_code(cls, code: int) -> "Tile":
        """从数字编码创建麻将牌"""
        if 1 <= code <= 9:
            return cls(type=TileType.WAN, value=code)
        elif 11 <= code <= 19:
            return cls(type=TileType.TIAO, value=code - 10)
        elif 21 <= code <= 29:
            return cls(type=TileType.TONG, value=code - 20)
        elif 31 <= code <= 37:
            return cls(type=TileType.ZI, value=code - 30)
        else:
            raise ValueError(f"Invalid tile code: {code}")
    
    def to_code(self) -> int:
        """转换为数字编码"""
        if self.type == TileType.WAN:
            return self.value
        elif self.type == TileType.TIAO:
            return self.value + 10
        elif self.type == TileType.TONG:
            return self.value + 20
        elif self.type == TileType.ZI:
            return self.value + 30


class MeldType(Enum):
    """组合类型"""
    PENG = "peng"    # 碰
    GANG = "gang"    # 杠
    CHI = "chi"      # 吃


class Meld(BaseModel):
    """牌组合（碰、杠、吃）"""
    type: MeldType
    tiles: List[Tile]
    exposed: bool = True  # 是否明牌


class HandTiles(BaseModel):
    """手牌"""
    tiles: List[Tile] = []
    melds: List[Meld] = []
    
    def tile_count(self) -> int:
        """总牌数"""
        return len(self.tiles) + sum(len(meld.tiles) for meld in self.melds)


class PlayerAction(BaseModel):
    """玩家动作"""
    player_id: int
    action_type: str  # "draw", "discard", "peng", "gang", "chi"
    tiles: List[Tile]
    timestamp: Optional[float] = None


class GameState(BaseModel):
    """游戏状态"""
    player_hands: Dict[int, HandTiles] = {}  # 玩家手牌
    discarded_tiles: List[Tile] = []         # 弃牌池
    remaining_tiles: List[Tile] = []         # 剩余牌
    current_player: int = 0
    actions_history: List[PlayerAction] = []
    
    def get_visible_tiles(self) -> List[Tile]:
        """获取所有可见的牌"""
        visible = []
        visible.extend(self.discarded_tiles)
        
        for hand in self.player_hands.values():
            for meld in hand.melds:
                if meld.exposed:
                    visible.extend(meld.tiles)
        
        return visible
    
    def calculate_remaining_tiles(self) -> Dict[int, int]:
        """计算剩余牌数量"""
        # 初始化所有牌的数量（每种牌4张）
        all_tiles = {}
        for code in range(1, 10):  # 万
            all_tiles[code] = 4
        for code in range(11, 20):  # 条
            all_tiles[code] = 4
        for code in range(21, 30):  # 筒
            all_tiles[code] = 4
        for code in range(31, 38):  # 字牌
            all_tiles[code] = 4
        
        # 减去已知的牌
        visible_tiles = self.get_visible_tiles()
        for hand in self.player_hands.values():
            visible_tiles.extend(hand.tiles)
        
        for tile in visible_tiles:
            code = tile.to_code()
            if code in all_tiles:
                all_tiles[code] = max(0, all_tiles[code] - 1)
        
        return all_tiles


class AnalysisResult(BaseModel):
    """分析结果"""
    recommended_discard: Optional[Tile] = None
    discard_scores: Dict[str, float] = {}
    listen_tiles: List[Tile] = []
    win_probability: float = 0.0
    remaining_tiles_count: Dict[int, int] = {}
    suggestions: List[str] = []


class GameRequest(BaseModel):
    """游戏请求"""
    game_state: GameState
    target_player: int = 0  # 目标玩家ID（通常是自己）


class GameResponse(BaseModel):
    """游戏响应"""
    success: bool
    analysis: Optional[AnalysisResult] = None
    message: str = "" 

File: frontend/src/types/mahjong.ts
export enum TileType {
  WAN = "wan",      // 万
  TIAO = "tiao",    // 条
  TONG = "tong",    // 筒
  ZI = "zi"         // 字牌
}

export interface Tile {
  type: TileType;
  value: number;
  id?: string;      // 添加牌的唯一标识
}

export enum MeldType {
  PENG = "peng",    // 碰
  GANG = "gang",    // 杠
  CHI = "chi"       // 吃
}

export enum GangType {
  MING_GANG = "ming_gang",  // 明杠（直杠）
  AN_GANG = "an_gang",      // 暗杠
  JIA_GANG = "jia_gang"     // 加杠
}

export interface Meld {
  id?: string;      // 添加唯一标识
  type: MeldType;
  tiles: Tile[];
  exposed: boolean;
  gang_type?: GangType; // 杠牌类型（仅当type为GANG时有效）
  source_player?: number;        // 直杠时：被杠牌的来源玩家ID
  original_peng_id?: string;     // 加杠时：原碰牌的ID
  timestamp?: number;            // 操作时间戳
}

export interface HandTiles {
  tiles: Tile[];
  melds: Meld[];
}

export interface PlayerAction {
  player_id: number;
  action_type: string;
  tiles: Tile[];
  timestamp?: number;
}

export interface GameState {
  player_hands: { [key: number]: HandTiles };
  discarded_tiles: Tile[];
  player_discarded_tiles?: { [key: number]: Tile[] };
  remaining_tiles: Tile[];
  current_player: number;
  actions_history: PlayerAction[];
}

export interface AnalysisResult {
  recommended_discard?: Tile;
  discard_scores: { [key: string]: number };
  listen_tiles: Tile[];
  win_probability: number;
  remaining_tiles_count: { [key: number]: number };
  suggestions: string[];
}

export interface GameRequest {
  game_state: GameState;
  target_player: number;
}

export interface GameResponse {
  success: boolean;
  analysis?: AnalysisResult;
  message: string;
}

export interface TileInfo {
  code: number;
  type: string;
  value: number;
  display: string;
}

// 输入模式枚举
export enum InputMode {
  ADD_TO_HAND = "add_to_hand",
  DISCARD = "discard", 
  PENG = "peng",
  MING_GANG = "ming_gang",
  AN_GANG = "an_gang"
}

// 工具函数
export const createTile = (type: TileType, value: number): Tile => ({
  type,
  value
});

export const tileToCode = (tile: Tile): number => {
  switch (tile.type) {
    case TileType.WAN:
      return tile.value;
    case TileType.TIAO:
      return tile.value + 10;
    case TileType.TONG:
      return tile.value + 20;
    case TileType.ZI:
      return tile.value + 30;
    default:
      throw new Error(`Invalid tile type: ${tile.type}`);
  }
};

export const codeToTile = (code: number): Tile => {
  if (1 <= code && code <= 9) {
    return { type: TileType.WAN, value: code };
  } else if (11 <= code && code <= 19) {
    return { type: TileType.TIAO, value: code - 10 };
  } else if (21 <= code && code <= 29) {
    return { type: TileType.TONG, value: code - 20 };
  } else if (31 <= code && code <= 37) {
    return { type: TileType.ZI, value: code - 30 };
  } else {
    throw new Error(`Invalid tile code: ${code}`);
  }
};

export const tileToString = (tile: Tile): string => {
  const typeMap = {
    [TileType.WAN]: "万",
    [TileType.TIAO]: "条", 
    [TileType.TONG]: "筒",
    [TileType.ZI]: ["", "东", "南", "西", "北", "中", "发", "白"][tile.value] || ""
  };
  
  if (tile.type === TileType.ZI) {
    return typeMap[tile.type];
  } else {
    return `${tile.value}${typeMap[tile.type]}`;
  }
};

export const tilesEqual = (tile1: Tile, tile2: Tile): boolean => {
  return tile1.type === tile2.type && tile1.value === tile2.value;
};

// 计算剩余牌数
export const calculateRemainingTiles = (gameState: GameState): number => {
  const totalTiles = 144; // 标准麻将总牌数
  
  // 计算已使用的牌数
  let usedTiles = 0;
  
  // 计算所有玩家手牌数量
  Object.values(gameState.player_hands).forEach(hand => {
    usedTiles += hand.tiles.length;
    // 计算碰牌杠牌数量
    hand.melds.forEach(meld => {
      usedTiles += meld.tiles.length;
    });
  });
  
  // 计算弃牌数量
  usedTiles += gameState.discarded_tiles.length;
  
  return Math.max(0, totalTiles - usedTiles);
}; 